src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+            _cancellationTokenSource.Dispose();
+        }
+
+        private async ValueTask<DocumentSymbolDataModel?> GetDocumentSymbolAsync(CancellationToken cancellationToken)
+        {
+            _threadingContext.ThrowIfNotOnBackgroundThread();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var textBuffer = _textBuffer;
+            var filePath = _textBuffer.GetRelatedDocuments().FirstOrDefault(static d => d.FilePath is not null)?.FilePath;
+            if (filePath is null)
+            {
+                // text buffer is not saved to disk
+                // LSP does not support calls without URIs
+                // and Visual Studio does not have a URI concept other than the file path.
+                return null;
+            }
+
+            // Obtain the LSP response and text snapshot used.
+            var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                textBuffer, _languageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+            // If there is no matching LSP server registered the client will return null here - e.g. wrong content type on the buffer, the
+            // server totally failed to start, server doesn't support the right capabilities. For C# we might know it's a bug if we get a null
+            // response here, but we don't know that in general for all languages.
+            // see "Microsoft.CodeAnalysis.Editor.Implementation.LanguageClient.AlwaysActivateInProcLanguageClient" for the list of content types we register for.
+            // At this time the expected list is C#, Visual Basic, and F#
+            if (response is null)
+            {
+                return null;
+            }
+
+            var responseBody = response.Value.response.ToObject<DocumentSymbol[]>();
+            // It would be a bug in the LSP server implementation if we get back a null result here.
+            Assumes.NotNull(responseBody);
+
+            var model = DocumentOutlineHelper.CreateDocumentSymbolDataModel(responseBody, response.Value.snapshot);
+
+            _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: true));
+
+            return model;
+        }
+
+        private void EnqueueFilter(string? newText)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(newText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueSelectTreeNode(CaretPosition caretPoint)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, caretPoint, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueExpandOrCollapse(ExpansionOption option)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, option, DataUpdated: false));
```

Comment:
are lines 186/192/195/198 the only places you make the ViewModelStateDataChange's?

Reply:
Yes